### PR TITLE
fix(filetype): avoid recursive FileType autocmds

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

**Comment by clason** [comment]
Are you sure those _should_ be avoided? Test failure looks related.

**Comment by lewis6991** [comment]
Yep, when I change `get_lang` to use `vim.filetype.get_option(ft, 'parser')`, nvim-treesitter ran in recursion hell because it calls that function in a `FileType` autocmd.

File: src/nvim/autocmd.c [hunk 1]
```diff
@@ -2758,3 +2758,32 @@ void do_autocmd_focusgained(bool gained)
...
+  // Reset the secure flag, since the value of 'filetype' has
+  // been checked to be safe.
+  secure = 0;
+
+  recursive = true;
+  did_filetype = true;
+  // Only pass true for "force" when it is true or
+  // used recursively, to avoid endless recurrence.
+  apply_autocmds(EVENT_FILETYPE, buf->b_p_ft, buf->b_fname, force, buf);
+  recursive = false;
```
Comment (zeertzjq) [inline src/nvim/autocmd.c:2782]: This doesn't look right. With `force = true` and recursive autocommand this is set to `false` too early.

Comment (lewis6991) [reply src/nvim/autocmd.c:2782]: Why? Shouldn't matter as long as it's after the autocmd call. It's set in the same place it was before with `ft_recursive`.

Comment (lewis6991) [reply src/nvim/autocmd.c:2782]: Ah I see what you mean now. Now it makes why the variable was an int before.

Thanks.

File: src/nvim/api/options.c
Comment (devilshezi0) [inline src/nvim/api/options.c:1]: src/nvim/api/options.c

**Comment by devilshezi0** [comment]
https://github.com/neovim/neovim/pull/22813#discussion_r1153973028

**Comment by devilshezi0** [comment]
https://github.com/neovim/neovim/pull/22813#discussion_r1153973028

**Pull request merged**